MemoryGauge app: log memory in use for workers in SHM and RRD

eugeneia opened the pull request:
- Adds a SHM type “gauge“ for double precision floating point values
- Adds an app that periodically writes `collectgarbage('count')` (kbytes in use) to a SHM gauge (and configure it for lwaftr workers)
- Extend ptree manager to aggregate gauge RRDs

The result is that lwaftr instances have RRDs like `apps/memory_monitor/kbytes_in_use.rrd` which you can plot with `rrdtool graph` to get graphs like the following:

![mem_gauge](https://user-images.githubusercontent.com/4933566/64945359-a6547380-d870-11e9-8418-a840489f2380.png)

eugeneia committed:
lib.rrd: fix a bug due to implicit casting

eugeneia committed:
lwaftr/apps: add memory_info.MemoryLog app to log mem in use to RRD

eugeneia committed:
lwaftr: configure MemoryLog app for lwaftr instances

eugeneia committed:
lwaftr: replace MemoryLog w/ MemoryCounter (ptree man. gen. rrd)

eugeneia committed:
lib.gauge: add SHM object type for gauge values

eugeneia committed:
apps.lwaftr.memory_info: use lib.gauge to expose a gauge SHM object

eugeneia committed:
lib.pree: support aggregating and RRD generation for SHM gauges

wingo commented on src/lib/rrd.lua:
```diff
@@ -400,8 +400,9 @@ end
    -- This condition comes from upstream rrdtool.
-   if dt < heartbeat_interval and pdp.unknown_count < secs_per_pdp/2 then
-      return pdp.diff / tonumber(secs_per_pdp - pdp.unknown_count)
+   local unknown = tonumber(pdp.unknown_count)
```
Pretty sure that as pdp is an FFI type (coming from one of the pdp_prep instances), pdp.unknown_count should always be a uint64_t.  WDYT?

wingo replied:
Ah I have been doing JS for too long :P  I see that this is a cast to a double, not a cast of not-a-number to a number.  I guess `<` wasn't working right comparing uint64 and doubles?  Was it casting doubles to uint64 or something?  Anyway cool :)

eugeneia replied:
I think the issue was that it is cast to an int which is fine for counter type RRDs (where the PDPs are integer values) but broke gauge type RRDs (PDPs are doubles.) Effectively IIRC the problem was truncation to zero for gauges with change rates of zero-point-something per second.

wingo commented on src/lib/gauge.lua:
```diff
@@ -0,0 +1,48 @@
+local ffi = require("ffi")
+local shm = require("core.shm")
+
+-- SHM object type for gauges (double precision float values).
```
Excellent :-)

wingo commented:
Looks great, thanks.  Two questions: one, do you think this makes sense as an app?  To me it's more like a timer, like the ingress drop monitor, or the ptree manager stuff.  WDYT?  Would be a small change but if you feel it's better as an app I'm happy to hear that argument.  Second, do you have a use example? :-)  Really cool hack, looks quite useful!

eugeneia commented:
> Looks great, thanks. Two questions: one, do you think this makes sense as an app? To me it's more like a timer, like the ingress drop monitor, or the ptree manager stuff. WDYT? Would be a small change but if you feel it's better as an app I'm happy to hear that argument. Second, do you have a use example? :-) Really cool hack, looks quite useful!

Doesn’t need to be an app, and I agree its somewhat counter intuitive. The reason it is an app are for convenience:
- convenient reuse of the app SHM frame mechanic
- can be easily configured per process using the ptree infrastructure

eugeneia commented on src/program/lwaftr/setup.lua:
```diff
@@ -141,6 +142,9 @@ function lwaftr_app(c, conf)
 
+   -- Add MemoryLog app.
+   config.app(c, "memory_monitor", memory_info.MemoryGauge)
+
```
@wingo For a usage example, you mean like this? :-)

I see this as sort of a benefit of it being an app. Super painless configuration. Hack? Maybe.

eugeneia replied:
I.e., one-line change for a memory gauge and RRD for every lwaftr process.

wingo replied:
I'm definitely down with "everybody gets a counter" :-)  I would add this in `lib.ptree` though at least in `lib.ptree.worker`, so that all ptree apps get this counter :)  WDYT?

wingo replied:
i.e. adding a timer there, as `lib.scheduling` adds an ingress drop monitor.  But if you think that doesn't work, that's ok too.  LMK.

wingo replied:
Fix in https://github.com/eugeneia/snabb/pull/16 .  wdyt @eugeneia ?

eugeneia replied:
Merged, LGTM!

wingo committed:
Make memory monitor a timer instead of an app

The ptree worker will automatically install a heap size monitor,
which writes to engine/heap_size_bytes.{gauge,rrd}.

eugeneia committed:
Merge pull request #16 from Igalia/lwaftr-memory-info

Make memory monitor a timer instead of an app

wingo commented:
Thanks!
